### ISSUE TITLE
sound.play() on 1.9.1 does not "stop"

### ISSUE THREAD

**Issue by davelawr** [kivy/kivy#3845]
```python
>>> sound = SoundLoader.load('sound.wav')
>>> sound.play()
>>> sound.play()
```

The first play plays the sound, but it then doesn't enter 'stop' state when it finishes and the second play does not re-play the sound. a sound.stop() or sound.seek(0) is needed.

Seen on 1.9.1 / rpi release as well and Mint (ubuntu 14.04 base)

**Comment by Kovak** [comment]
Are you sure this is not because the first play is still in the middle of playing?

If you want to play 2 instances of one sound simultaneously you must load 2 sound objects.

**Comment by davelawr** [comment]
No,

sound.play() plays the sound, and then the sound ends, but it does not 
appear to enter the 'stop' state.

On 14/12/15 21:43, Kovak wrote:
>
> Are you sure this is not because the first play is still in the middle 
> of playing?
>
> If you want to play 2 instances of one sound simultaneously you must 
> load 2 sound objects.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/kivy/kivy/issues/3845#issuecomment-164568899>.
>

-- 
.  Dave Lawrence  .  Tel: +44 1628 491960  .  Net:  dl@alphawave.net  .
|  Alphawave Ltd  |  Tel: +44 7768 224315  |  Web: www.alphawave.net  |

**Comment by Kovak** [comment]
You should perhaps share an example of your actual use case. The tiny snippet you provided is contradictory to your account of what is occurring. In the snippet you provided you are definitely trying to play 1 sound twice, simultaneously.

**Comment by davelawr** [comment]
Hard to give an example as it's entirely audio, but here is the sequence 
of events:

 >>> sound.play()
  (( sound sample plays to end of sample, i.e. ~1s )))
 >>> print sound.state
'play'
 >>> sound.play()
  (( nothing ))
 >>> sound.stop()
 >>> sound.play()
   (( sample plays a second time)

The same sequence on 1.8.0 / ubuntu plays the sample a second time.

Please see thread where I first made the query. ZenCode onfirmed same 
operation on 1.8.0 / mint

https://groups.google.com/forum/#!topic/kivy-users/r3OdwHaZbJg

Thanks,

On 14/12/15 22:32, Kovak wrote:
>
> You should perhaps share an example of your actual use case. The tiny 
> snippet you provided is contradictory to your account of what is 
> occurring. In the snippet you provided you are definitely trying to 
> play 1 sound twice, simultaneously.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/kivy/kivy/issues/3845#issuecomment-164581043>.
>

-- 
.  Dave Lawrence  .  Tel: +44 1628 491960  .  Net:  dl@alphawave.net  .
|  Alphawave Ltd  |  Tel: +44 7768 224315  |  Web: www.alphawave.net  |

**Comment by Kovak** [comment]
Which sound provider is being used? Can you post the startup log output for your kivy application?

This behavior is not exhibited on current master with the SDL2 audio provider.

Also it is still desirable to have a precise example with an actual sound. This would allow further inspection as to the root cause of your problem.

**Comment by davelawr** [comment]
Here's an example using one of the audio samples from kivy-examples

#!/usr/bin/python
from kivy.core.audio import SoundLoader
import time
sound = 
SoundLoader.load("/usr/local/share/kivy-examples/audio/12922_sweet_trip_mm_kwik_mod_07.wav")
print "1) state %s length %r" % (sound.state, sound.length)
sound.play()
print "2) state %s" % sound.state
time.sleep(2)
print "3) state %s" % sound.state
sound.play()
print "4) state %s" % sound.state

And here's the output from a run:

$ python ./example.py
[INFO   ] [Logger      ] Record log in 
/home/sysop/.kivy/logs/kivy_15-12-15_7.txt
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_gif, img_sdl2, 
img_pil (img_ffpyplayer ignored)
[INFO   ] [Kivy        ] v1.9.1-dev
[INFO   ] [Python      ] v2.7.9 (default, Mar  8 2015, 00:52:26)
[GCC 4.9.2]
[INFO   ] [AudioGstplayer] Using Gstreamer 1.4.4.0
[INFO   ] [Audio       ] Providers: audio_gstplayer, audio_sdl2 
(audio_ffpyplayer ignored)
1) state stop length 0.116780048
2) state play
3) state play
4) state play

the on_stop event is not triggered either.

let me know if you need anything else.

On 15/12/15 00:44, Kovak wrote:
>
> Which sound provider is being used? Can you post the startup log 
> output for your kivy application?
>
> This behavior is not exhibited on current master with the SDL2 audio 
> provider.
>
> Also it is still desirable to have a precise example with an actual 
> sound. This would allow further inspection as to the root cause of 
> your problem.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/kivy/kivy/issues/3845#issuecomment-164605579>.
>

-- 
.  Dave Lawrence  .  Tel: +44 1628 491960  .  Net:  dl@alphawave.net  .
|  Alphawave Ltd  |  Tel: +44 7768 224315  |  Web: www.alphawave.net  |

**Comment by Kovak** [comment]
If this is indeed how you are using sound, it is because you do not even have an event loop running and also the sleep blocks the main thread.

Using the gstreamer provider and ensuring the event loop is running and I am not blocking it, I get the on_stop being called and 2 sounds playing.

```python
import kivy
from kivy.app import App
from kivy.uix.widget import Widget
from kivy.clock import Clock
from kivy.core.audio import SoundLoader
from functools import partial

class YourAppNameApp(App):
    def build(self):
        return Widget()

    def on_start(self):
        sound = SoundLoader.load("testsound.wav") #insert a sound here
        sound.bind(on_stop=self.stop_event)
        Clock.schedule_once(partial(self.play_sound, sound))
        Clock.schedule_once(partial(self.play_sound, sound), 2.0)

    def play_sound(self, sound, dt):
        sound.play()

    def stop_event(self, sound):
        print(sound, 'stopping')


if __name__ == '__main__':
    YourAppNameApp().run()
```

**Comment by davelawr** [comment]
Hi,

Your example gives the same effect when running rpi/1.9.1 - the first sound plays, but then nothing else.
I have asked if ZenCode can test too as they were able to confirm my original findings.

Both yours and my original "manual" tests work fine on ubuntu/1.8.0 with no App and no event loop.

Thanks,

**Comment by Zen-CODE** [comment]
Not sure how it worked previously, perhaps it was a different provider?

All I can say is that when I test on Mint 17 (14.04), it behaves consistently as Kovac describes regarding the event loop. I have added that to the docs (suggestion/corrections welcome) to try and explain that better.

https://github.com/kivy/kivy/commit/8e02046b2dc1cc0d56501fdaa2b1eece47c0ac84

Otherwise, you say you still an issue on rpi? Have you tested on a Android device? Using the Kivy launcher should make this easy?

**Comment by Zen-CODE** [comment]
Of course, easy is relative :-). If it's not, just say and I can test on my phone...

**Comment by akshayaurora** [comment]
I can test on rpi.

**Comment by akshayaurora** [comment]
Issue confirmed on api

**Comment by mgc8** [comment]
I am seeing the exact same problem with Kivy 1.11.1 and Linux (Ubuntu 18.04 if it matters). The code above behaves as described, it plays once and then "dies"; my own (quite different) code does the same. Is this a known issue?
